pareto/strategic_water_management/strategic_produced_water_optimization.py

```diff
@@ -4069,6 +4159,575 @@ def DisposalDestinationDeliveriesRule(model, k, t):
+        model.s_W,
+        model.s_T,
+        rule=CompletionsPadIntermediateWaterQuality,
+        doc="Completions pad water quality",
+    )
+
+    def CompletionsPadWaterQuality(b, p, w, t):
+        return (
+            b.parent_block().v_F_PadStorageOut[p, t] * b.v_Q[p + "-storage", w, t]
+            + b.parent_block().v_F_CompletionsWater[p, t]
+            * b.v_Q[p + "-intermediate", w, t]
+            == b.v_Q[p, w, t] * b.parent_block().p_gamma_Completions[p, t]
+        )
+
+    model.quality.CompletionsPadWaterQuality = Constraint(
+        model.s_CP,
+        model.s_W,
+        model.s_T,
+        rule=CompletionsPadWaterQuality,
+        doc="Completions pad water quality",
+    )
+    # endregion
+
+    # region Completion Pad Storage
+    def CompletionsPadStorageWaterQuality(b, p, w, t):
+        if t == b.parent_block().s_T.first():
+            return b.p_xi_PadStorage[p, w] * b.parent_block().p_lambda_PadStorage[
+                p
+            ] * +b.v_Q[p + "-intermediate", w, t] * b.parent_block().v_F_PadStorageIn[
+                p, t
+            ] == b.v_Q[
+                p + "-storage", w, t
+            ] * (
+                b.parent_block().v_L_PadStorage[p, t]
+                + b.parent_block().v_F_PadStorageOut[p, t]
+            )
+        else:
+            return b.v_Q[
+                p + "-storage", w, b.parent_block().s_T.prev(t)
+            ] * b.parent_block().v_L_PadStorage[
+                p, b.parent_block().s_T.prev(t)
+            ] + b.v_Q[
+                p + "-intermediate", w, t
+            ] * b.parent_block().v_F_PadStorageIn[
+                p, t
+            ] == b.v_Q[
+                p + "-storage", w, t
+            ] * (
+                b.parent_block().v_L_PadStorage[p, t]
+                + b.parent_block().v_F_PadStorageOut[p, t]
+            )
+
+    model.quality.CompletionsPadStorageWaterQuality = Constraint(
+        model.s_CP,
+        model.s_W,
+        model.s_T,
+        rule=CompletionsPadStorageWaterQuality,
+        doc="Completions pad storage water quality",
+    )
+    # endregion
+
+    # Define Objective
+    def ObjectiveFunctionRule(b):
+        return b.v_X == sum(
+            sum(
+                sum(b.v_Q[p, w, t] for p in b.parent_block().s_P)
+                for w in b.parent_block().s_W
+            )
+            for t in b.parent_block().s_T
+        )
+
+    model.quality.ObjectiveFunction = Constraint(
+        rule=ObjectiveFunctionRule, doc="Objective function water quality"
+    )
+
+    model.quality.objective = Objective(
+        expr=model.quality.v_X, sense=minimize, doc="Objective function"
+    )
+
+    return model
+
+
+def postprocess_water_quality_calculation(model, df_parameters, df_sets, opt):
+    # Add water quality formulation to input solved model
+    water_quality_model = water_quality(model, df_parameters, df_sets)
+
+    # Calculate water quality. The following conditional is used to avoid errors when
```

Comment:
What kind of errors do we get with Gurobi?

Reply:
The call " opt.solve(water_quality_model.quality, tee=True)" resulted in an error on Andres and Miguel's laptop. I believe it was this error: AttributeError: 'ScalarBlock' object has no attribute 'solutions' (https://github.com/Pyomo/pyomo/issues/2130)

Adding save_results=False to the .solve() call fixes the error for Andres/Miguel, but it causes a ValueError on our laptops . I added this try except to address both cases.